apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/SizedBufferingConsumer.cs

```diff
@@ -55,50 +62,61 @@ public static SizedBufferingConsumer<T> Create(IConsumer<T> consumer, ISizer<T>
             => new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
 
         /// <inheritdoc />
-        public override void Receive(IEnumerable<T> items)
+        protected override void ReceiveWithSemaphoreHeld(IEnumerable<T> items)
         {
             GaxPreconditions.CheckNotNull(items, nameof(items));
-            Semaphore.Wait();
-            try
+            foreach (T item in items)
             {
-                foreach (T item in items)
+                _size += _sizer.GetSize(item);
+                _items.Add(item);
+                if (_size >= _bufferSize)
                 {
-                    _size += _sizer.GetSize(item);
-                    Items.Add(item);
-                    if (_size >= _bufferSize)
-                    {
-                        FlushWithSemaphoreHeld();
-                    }
+                    FlushWithSemaphoreHeld();
                 }
             }
-            finally
-            {
-                Semaphore.Release();
-            }
         }
 
         /// <inheritdoc />
-        public override async Task ReceiveAsync(
+        protected override Task ReceiveAsyncWithSemaphoreHeld(
             IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
         {
             GaxPreconditions.CheckNotNull(items, nameof(items));
-            await Semaphore.WaitAsync(cancellationToken);
-            try
+            foreach (T item in items)
             {
-                foreach (T item in items)
+                _size += _sizer.GetSize(item);
+                _items.Add(item);
+                if (_size >= _bufferSize)
                 {
-                    _size += _sizer.GetSize(item);
-                    Items.Add(item);
-                    if (_size >= _bufferSize)
-                    {
-                        await FlushAsyncWithSemaphoreHeld(cancellationToken);
-                    }
+                    return FlushAsyncWithSemaphoreHeld(cancellationToken);
```

Comment:
This doesn't look right. This will return after the first flush and potentially not receive all items. I would change this back to use `async` and `await` as before.